### PR TITLE
Reworks dockerConfig auth setup

### PR DESCRIPTION
**Pull request by 0sewa0**
# Description
The `NewDockerConfig` constructor will not setup the `Auths`, that is done in `SetupAuths` method.
This is a small optimisation so in case something needs the dockerConfig but requires no authentication, we will no longer force to use the default pull secret. (Also its nicer for testing)

## How can this be tested?
When you use `codeModulesImage` that points to a registry that doesn't need `customPullSecret`, it should work

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](/CONTRIBUTING.md)

### CODE REVIEW

File: src/dockerconfig/docker_config.go [hunk 1]
```diff
@@ -25,47 +25,30 @@ type DockerAuth struct {
...
+	if err != nil {
+		log.Info("failed to load pull secret", "dynakube", config.Dynakube.Name)
+		return errors.WithStack(err)
 	}
-	err := json.Unmarshal(config, &dockerConf)
+	dockerAuths, err := parseDockerAuthsFromSecret(&pullSecret)
 	if err != nil {
-		return nil, err
+		log.Info("failed to parse pull secret content", "dynakube", config.Dynakube.Name)
+		return err
```
Comment (luhi-DT) [inline src/dockerconfig/docker_config.go:48]:
```suggestion
		return errors.WithStack(err)
```

File: src/dockerconfig/docker_config.go [hunk 2]
```diff
@@ -76,15 +59,36 @@ func (config *DockerConfig) SaveCustomCAs(
...
-		return err
+		return errors.WithStack(err)
 	}
 	config.TrustedCertsPath = path
 	return nil
 }
+
+func parseDockerAuthsFromSecret(secret *corev1.Secret) (map[string]DockerAuth, error) {
+	if secret == nil {
+		return nil, errors.New("given secret is nil")
```
Comment (luhi-DT) [inline src/dockerconfig/docker_config.go:77]: could this be a more specific error message ? could not generate docker auths from empty secret

File: src/controllers/csi/provisioner/agent.go [hunk 3]
```diff
@@ -100,9 +100,12 @@ func getUrlProperties(version string) *url.Properties {
 }
 
 func setupImageInstaller(ctx context.Context, fs afero.Fs, pathResolver metadata.PathResolver, apiReader client.Reader, certPath, digest string, dynakube *dynatracev1beta1.DynaKube) (installer.Installer, error) {
-	dockerConfig, err := dockerconfig.NewDockerConfig(ctx, apiReader, *dynakube)
-	if err != nil {
-		return nil, err
+	dockerConfig := dockerconfig.NewDockerConfig(apiReader, *dynakube)
+	if dynakube.Spec.CustomPullSecret != "" {
```
Comment (luhi-DT) [inline src/controllers/csi/provisioner/agent.go:104]: maybe refactor in a seperate function like `hasCustomPullSecret`

File: src/dockerconfig/docker_config.go [hunk 4]
```diff
@@ -25,47 +25,30 @@ type DockerAuth struct {
...
-		return nil, err
-	}
-
+func NewDockerConfig(apiReader client.Reader, dynakube dynatracev1beta1.DynaKube) *DockerConfig {
 	dockerConfig := DockerConfig{
 		ApiReader:     apiReader,
+		Auths:         make(map[string]DockerAuth),
 		Dynakube:      &dynakube,
-		Auths:         dockerAuths,
 		SkipCertCheck: dynakube.Spec.SkipCertCheck,
```
Comment (luhi-DT) [inline src/dockerconfig/docker_config.go:33]: I dont think we would need SkipCertCheck to be part of the struct, as we have the dynakube as a whole anyway

**Review by luhi-DT — CHANGES_REQUESTED** [review]

File: src/dockerconfig/docker_config.go (hunk 1 above)
Comment (meik99) [reply src/dockerconfig/docker_config.go:48]: that error comes from an internal function, `err` is fine

File: src/controllers/csi/provisioner/agent.go (hunk 3 above)
Comment (0sewa0) [reply src/controllers/csi/provisioner/agent.go:104]: I don't think its necessary, its a higher level property

`dynakube.Spec.TrustedCAs != ""` should also need to be replaced to remain consistent

File: src/dockerconfig/docker_config.go (hunk 4 above)
Comment (0sewa0) [reply src/dockerconfig/docker_config.go:33]: made it a method, so we don't have to dig for it 👍

File: src/dockerconfig/docker_config.go (hunk 1 above)
Comment (luhi-DT) [reply src/dockerconfig/docker_config.go:46]: makes sense

**Review by luhi-DT — APPROVED** [review]

**Pull request merged**